File: ios/Core/Domain/Sources/CoreDomain/MusicLibraryModels.swift
import Foundation

// MARK: - Semantic identifiers

public struct ProviderUserId: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }
}

public struct PlaylistKind: RawRepresentable, Hashable, Sendable {
    public let rawValue: Int64

    public init(rawValue: Int64) {
        self.rawValue = rawValue
    }
}

public struct PlaylistUuid: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }
}

public struct TrackId: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }
}

public struct AlbumId: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }
}

public struct PlaylistId: Hashable, Sendable {
    public let ownerId: ProviderUserId
    public let kind: PlaylistKind

    public init(ownerId: ProviderUserId, kind: PlaylistKind) {
        self.ownerId = ownerId
        self.kind = kind
    }
}

public struct TrackRef: Hashable, Sendable {
    public let trackId: TrackId
    public let albumId: AlbumId?

    public init(trackId: TrackId, albumId: AlbumId?) {
        self.trackId = trackId
        self.albumId = albumId
    }
}

// MARK: - Enums

public enum MusicProviderId: Sendable {
    case yandexMusic
}

public enum PlaylistVisibility: Sendable {
    case `public`
    case `private`
}

// MARK: - Service availability

public struct MusicServiceAvailability: Sendable {
    public let isAvailable: Bool
    public let region: Int?
    public let permissions: Set<String>

    public init(isAvailable: Bool, region: Int?, permissions: Set<String>) {
        self.isAvailable = isAvailable
        self.region = region
        self.permissions = permissions
    }
}

// MARK: - Playlist models

public struct PlaylistSummary: Sendable {
    public let id: PlaylistId
    public let provider: MusicProviderId
    public let playlistUuid: PlaylistUuid?
    public let title: String
    public let ownerName: String?
    public let coverUriTemplate: String?
    public let trackCount: Int
    public let durationMs: Int64?
    public let isAvailable: Bool
    public let isCollective: Bool
    public let visibility: PlaylistVisibility?

    public init(
        id: PlaylistId,
        provider: MusicProviderId,
        playlistUuid: PlaylistUuid?,
        title: String,
        ownerName: String?,
        coverUriTemplate: String?,
        trackCount: Int,
        durationMs: Int64?,
        isAvailable: Bool,
        isCollective: Bool,
        visibility: PlaylistVisibility?
    ) {
        self.id = id
        self.provider = provider
        self.playlistUuid = playlistUuid
        self.title = title
        self.ownerName = ownerName
        self.coverUriTemplate = coverUriTemplate
        self.trackCount = trackCount
        self.durationMs = durationMs
        self.isAvailable = isAvailable
        self.isCollective = isCollective
        self.visibility = visibility
    }
}

public struct Playlist: Sendable {
    public let summary: PlaylistSummary
    public let revision: Int64?
    public let snapshot: Int64?
    public let likesCount: Int?
    public let tracks: [PlaylistTrackEntry]

    public init(
        summary: PlaylistSummary,
        revision: Int64?,
        snapshot: Int64?,
        likesCount: Int?,
        tracks: [PlaylistTrackEntry]
    ) {
        self.summary = summary
        self.revision = revision
        self.snapshot = snapshot
        self.likesCount = likesCount
        self.tracks = tracks
    }
}

public struct PlaylistTrackEntry: Sendable {
    public let position: Int
    public let addedAt: String?
    public let originalIndex: Int?
    public let originalShuffleIndex: Int?
    public let isRecent: Bool?
    public let trackRef: TrackRef
    public let track: Track?

    public init(
        position: Int,
        addedAt: String?,
        originalIndex: Int?,
        originalShuffleIndex: Int?,
        isRecent: Bool?,
        trackRef: TrackRef,
        track: Track?
    ) {
        self.position = position
        self.addedAt = addedAt
        self.originalIndex = originalIndex
        self.originalShuffleIndex = originalShuffleIndex
        self.isRecent = isRecent
        self.trackRef = trackRef
        self.track = track
    }
}

// MARK: - Saved tracks

public struct SavedTracks: Sendable {
    public let ownerId: ProviderUserId
    public let revision: Int64?
    public let tracks: [SavedTrackEntry]

    public init(ownerId: ProviderUserId, revision: Int64?, tracks: [SavedTrackEntry]) {
        self.ownerId = ownerId
        self.revision = revision
        self.tracks = tracks
    }
}

public struct SavedTrackEntry: Sendable {
    public let position: Int
    public let addedAt: String?
    public let trackRef: TrackRef
    public let track: TrackPreview?

    public init(position: Int, addedAt: String?, trackRef: TrackRef, track: TrackPreview?) {
        self.position = position
        self.addedAt = addedAt
        self.trackRef = trackRef
        self.track = track
    }
}

public enum SavedTracksResult: Sendable {
    case available(SavedTracks)
    case privateLibrary
}

// MARK: - Track models

public struct TrackPreview: Sendable {
    public let ref: TrackRef
    public let title: String
    public let artists: [ArtistPreview]
    public let durationMs: Int64?
    public let coverUriTemplate: String?
    public let isAvailable: Bool

    public init(
        ref: TrackRef,
        title: String,
        artists: [ArtistPreview],
        durationMs: Int64?,
        coverUriTemplate: String?,
        isAvailable: Bool
    ) {
        self.ref = ref
        self.title = title
        self.artists = artists
        self.durationMs = durationMs
        self.coverUriTemplate = coverUriTemplate
        self.isAvailable = isAvailable
    }
}

public struct Track: Sendable {
    public let preview: TrackPreview
    public let lyricsAvailable: Bool
    public let isAvailableForPremium: Bool
    public let isAvailableWithoutPermission: Bool

    public init(
        preview: TrackPreview,
        lyricsAvailable: Bool,
        isAvailableForPremium: Bool,
        isAvailableWithoutPermission: Bool
    ) {
        self.preview = preview
        self.lyricsAvailable = lyricsAvailable
        self.isAvailableForPremium = isAvailableForPremium
        self.isAvailableWithoutPermission = isAvailableWithoutPermission
    }
}

public struct ArtistPreview: Sendable {
    public let id: String
    public let name: String

    public init(id: String, name: String) {
        self.id = id
        self.name = name
    }
}

// MARK: - Errors

public enum MusicLibraryError: Error, Equatable, Sendable {
    case unauthorized
    case serviceUnavailable(description: String?)
    case providerError(code: String, description: String?)
    case invalidResponse(description: String)
    case networkFailure(reason: String)
}

File: android/core/domain/src/main/kotlin/com/mplayeraudio/core/domain/musiclibrary/MusicLibraryModels.kt
package com.mplayeraudio.core.domain.musiclibrary

enum class MusicProviderId {
    Device,
    YandexMusic,
    UserPlaylists,
}

@JvmInline
value class ProviderUserId(val value: String)

@JvmInline
value class PlaylistKind(val value: Long)

@JvmInline
value class PlaylistUuid(val value: String)

sealed interface TrackId {
    val stableKey: String
}

@JvmInline
value class YandexTrackId(val value: String) : TrackId {
    override val stableKey: String
        get() = value
}

@JvmInline
value class DeviceTrackId(val value: Long) : TrackId {
    override val stableKey: String
        get() = "device:$value"
}

@JvmInline
value class UserPlaylistTrackId(val value: Long) : TrackId {
    override val stableKey: String
        get() = "user-playlist:$value"
}

@JvmInline
value class AlbumId(val value: String)

data class PlaylistId(
    val ownerId: ProviderUserId,
    val kind: PlaylistKind,
)

data class PlaylistRef(
    val provider: MusicProviderId,
    val id: PlaylistId,
)

data class TrackRef(
    val trackId: TrackId,
    val albumId: AlbumId?,
)

enum class PlaylistVisibility {
    Public,
    Private,
}

enum class PlaylistRole {
    Favourites,
    Regular,
}

data class PlaylistSummary(
    val id: PlaylistId,
    val provider: MusicProviderId,
    val playlistUuid: PlaylistUuid?,
    val title: String,
    val ownerName: String?,
    val coverUriTemplate: String?,
    val trackCount: Int,
    val durationMs: Long?,
    val isAvailable: Boolean,
    val isCollective: Boolean,
    val visibility: PlaylistVisibility?,
    val role: PlaylistRole = PlaylistRole.Regular,
)

data class Playlist(
    val summary: PlaylistSummary,
    val revision: Long?,
    val snapshot: Long?,
    val likesCount: Int?,
    val tracks: List<PlaylistTrackEntry>,
)

data class PlaylistTrackEntry(
    val position: Int,
    val addedAt: String?,
    val originalIndex: Int?,
    val originalShuffleIndex: Int?,
    val isRecent: Boolean?,
    val trackRef: TrackRef,
    val track: Track?,
)

data class SavedTracks(
    val ownerId: ProviderUserId,
    val revision: Long?,
    val tracks: List<SavedTrackEntry>,
)

data class SavedTrackEntry(
    val position: Int,
    val addedAt: String?,
    val trackRef: TrackRef,
    val track: TrackPreview?,
)

data class TrackPreview(
    val ref: TrackRef,
    val title: String,
    val artists: List<ArtistPreview>,
    val durationMs: Long?,
    val coverUriTemplate: String?,
    val isAvailable: Boolean,
)

data class Track(
    val preview: TrackPreview,
    val lyricsAvailable: Boolean,
    val isAvailableForPremium: Boolean,
    val isAvailableWithoutPermission: Boolean,
)

data class ArtistPreview(
    val id: String,
    val name: String,
)

data class MusicServiceAvailability(
    val isAvailable: Boolean,
    val region: Int?,
    val permissions: Set<String>,
)

sealed interface SavedTracksResult {
    data class Available(
        val value: SavedTracks,
    ) : SavedTracksResult

    data object PrivateLibrary : SavedTracksResult
}

sealed interface MusicLibraryError

sealed class MusicLibraryException(
    message: String,
    cause: Throwable? = null,
) : Exception(message, cause), MusicLibraryError {

    class Unauthorized : MusicLibraryException(
        message = "Yandex Music access token is invalid or expired.",
    )

    data class ServiceUnavailable(
        val description: String?,
    ) : MusicLibraryException(
        message = description ?: "Yandex Music service is unavailable in the current environment.",
    )

    data class ProviderError(
        val code: String,
        val description: String?,
    ) : MusicLibraryException(
        message = description ?: code,
    )

    data class InvalidResponse(
        val description: String,
    ) : MusicLibraryException(
        message = description,
    )
}
